python/triton/compiler/code_generator.py

```diff
@@ -991,6 +991,9 @@ def visit_For(self, node):
 
             self.scf_stack.append(node)
             self.builder.set_insertion_point_to_start(for_op.get_body(0))
+            # reset local scope to not pick up local defs from the previous dry run.
+            self.lscope = liveins.copy()
+            self.local_defs = {}
```

Comment:
Then the `local_defs` on Line 1002 seems redundant.

https://github.com/triton-lang/triton/pull/4018/files#diff-a94d24c42ec01a10430ef002dabce1e275f194ead5123b418d518fbf92a6c4a0R1002

Note that `lscope` has another use on Line 1018.

https://github.com/triton-lang/triton/pull/4018/files#diff-a94d24c42ec01a10430ef002dabce1e275f194ead5123b418d518fbf92a6c4a0R1018

```
            for name in self.local_defs:
                if name in liveins:
                    yields.append(language.core._to_tensor(self.local_defs[name], self.builder))
            # create YieldOp
            if len(yields) > 0:
                self.builder.create_yield_op([y.handle for y in yields])
            for_op_region = for_op.get_body(0).get_parent()
            assert for_op_region.size() == 1, "We use SCF, so the loop body should only have one block"
            # update induction variable with actual value, and replace all uses
            self.builder.set_insertion_point_to_start(for_op.get_body(0))
            iv = for_op.get_induction_var()
            if negative_step:
                iv = self.builder.create_sub(ub, iv)
                iv = self.builder.create_add(iv, lb)
            self.lscope[node.target.id].handle.replace_all_uses_with(iv)
```

Reply:
Those uses are fine as `visit_compound_statement` will populate those array again. The problem is that we don't want the values to be set before calling it as the visiting goes in program order to figure when a variable is defined or not.

Reply:
I see it now. Thanks for explaining.